FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -0,0 +1,128 @@
+//
+//  CoreDataFeedStore.swift
+//  FeedStoreChallenge
+//
+//  Created by Tak Mazarura on 18/02/2021.
+//  Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+import Foundation
+import CoreData
+
+protocol CoreDataManager {
+	var managedObjectContext: NSManagedObjectContext { get set }
+	var persistentContainer: PersistentContainer { get set }
+}
+
+public final class CoreDataFeedStore: FeedStore {
+	private let container: PersistentContainer
+	private let context: NSManagedObjectContext
+	
+	public init () {
+		container = CoreDataFeedStore.setupContainer()
+		context = container.newBackgroundContext()
+	}
+	
+	static func setupContainer() -> PersistentContainer {
+		let container = PersistentContainer(name: "FeedStoreDataModel")
+		container.loadPersistentStores { description, error in
+			if let error = error {
+				fatalError("Unable to load persistent stores: \(error)")
+			}
+		}
+		return container
+	}
+}
+
+// MARK: - Create, Insert and Delete Functions
+extension CoreDataFeedStore {
+	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
+		perform { context in
+			do {
+				try FeedDB.fetch(in: context).map(context.delete)
+				try self.container.saveContext(backgroundContext: context)
+				completion(nil)
+			} catch {
+				completion(error)
+			}
+		}
+	}
+	
+	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
+		perform { context in
+			do {
+				try CoreDataFeedStore.createFeedDB(for: feed, timestamp: timestamp, context: context)
+				try self.container.saveContext(backgroundContext: context)
+				completion(nil)
+			} catch {
+				completion(error)
+			}
+		}
+	}
+	
+	public func retrieve(completion: @escaping RetrievalCompletion) {
+		perform { context in
+			do {
+				let feedDB = try FeedDB.fetch(in: context)
+				if let localFeedImages = feedDB?.localFeedImages, let timestamp = feedDB?.timestamp {
+					completion(.found(feed: localFeedImages, timestamp: timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
+	}
+}
+
+// MARK: Creation Methods
+extension CoreDataFeedStore {
+	@discardableResult
+	private static func createFeedDB(for localFeedImages: [LocalFeedImage], timestamp: Date, context: NSManagedObjectContext) throws -> FeedDB {
+		try FeedDB.fetch(in: context).map(context.delete)
+		
+		let feedDB = FeedDB(context: context, timestamp: timestamp)
+		feedDB.feedImageDBs = createFeedDBImages(from: localFeedImages, withContext: context)
+		feedDB.timestamp = timestamp
+		return feedDB
+	}
+	
+	private static func createFeedDBImages(from localFeedImages: [LocalFeedImage], withContext context: NSManagedObjectContext) -> NSOrderedSet {
+		NSOrderedSet(array: localFeedImages.map { localFeedImage in
+			let coreDataFeedImage = FeedImageDB(context: context)
+			coreDataFeedImage.id = localFeedImage.id
+			coreDataFeedImage.desc = localFeedImage.description ?? ""
```

Comment:
The database should not replace values with default values:

```suggestion
			coreDataFeedImage.desc = localFeedImage.description
```

Reply:
Updated! :)